File: helpers/macos/Sources/deskhop-helper/FileStore.swift
import DeskhopChannel
import Foundation

/*
 * Where files that arrive from the other computer are written (#56).
 *
 * Copying a file puts a *reference* on the clipboard, not its contents, so
 * pasting one that came over the link means there has to be a real file at a
 * real path to point at. This is that path — and because these are files
 * nobody asked to keep, it is a temporary directory that is swept when the
 * helper starts, rather than a folder in the user's own space that would slowly
 * fill up with things they never chose to save (#42, story 11).
 *
 * Emptied on **start** rather than on a timer or at exit, deliberately. A
 * helper that crashes never runs an exit path, and a timer would delete a file
 * the user is still working on.
 *
 * **The newest set survives that sweep**, and the reason is the crash. A file
 * reference on the pasteboard is a path, and a path outlives the process that
 * put it there — so a helper that crashed and was restarted by launchd,
 * seconds after files arrived, would delete the very files the user is about
 * to paste and leave a reference pointing at nothing. Keeping one set costs a
 * bounded amount of disk and is not accumulation: every run leaves at most one
 * behind.
 *
 * ---------------------------------------------------------------------------
 * NOTHING PARTIAL IS EVER LEFT BEHIND
 *
 * A transfer that fails delivers nothing at all — the transfer machine
 * discards an incomplete payload rather than handing it over — so this is only
 * ever called with every byte in hand. What can still fail is the writing, and
 * a set half-written to disk would put working references to truncated files
 * on the pasteboard. So a failure removes the whole directory and reports
 * nothing, which is the same rule one layer up.
 */
final class FileStore {
    /// Where a set of files landed, in the order they were named.
    struct Written {
        let directory: URL
        let urls: [URL]
    }

    var log: ((String) -> Void)?

    private let root: URL
    /// Distinguishes two sets that arrive inside the same second.
    private var sequence = 0

    init(root: URL = FileStore.defaultRoot) {
        self.root = root
    }

    static var defaultRoot: URL {
        URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
            .appendingPathComponent("deskhopplus", isDirectory: true)
    }

    /// Throw away what previous runs left, keeping only the newest set — see
    /// above for why that one is kept. Called once, at startup.
    func collectGarbage() {
        let manager = FileManager.default
        guard let names = try? manager.contentsOfDirectory(atPath: root.path) else { return }

        /* Newest by the counter in the name, which is what `write` builds it
           from: seconds since the epoch, then a per-run sequence. Sorting the
           names as numbers rather than as text keeps "10" after "9". */
        let newest = names.max { Self.ordering(of: $0) < Self.ordering(of: $1) }
        var removed = 0
        for name in names where name != newest {
            do {
                try manager.removeItem(at: root.appendingPathComponent(name))
                removed += 1
            } catch {
                log?("a set of files left by a previous run could not be removed: \(error)")
            }
        }
        if removed > 0 {
            log?("removed \(removed) set(s) of files left by a previous run; the newest was "
                 + "kept in case it is still on the pasteboard")
        }
    }

    /// The two numbers a set's directory name is built from, for ordering.
    private static func ordering(of name: String) -> (Int, Int) {
        let parts = name.split(separator: "-", maxSplits: 1)
        return (Int(parts.first ?? "") ?? 0, parts.count > 1 ? Int(parts[1]) ?? 0 : 0)
    }

    /*
     * Write a delivered set and return where each file landed.
     *
     * The names are the ones the shared core already accepted — it refuses any
     * that could name somewhere other than the directory they are joined to —
     * so nothing here re-examines them. What it does own is *collisions*:
     * cleaning can map two different names onto one, and two files copied from
     * different folders can share a name outright, and neither may quietly
     * overwrite the other.
     */
    func write(_ delivery: FileDelivery) -> Written? {
        sequence += 1
        let directory = root.appendingPathComponent(
            "\(Int(Date().timeIntervalSince1970))-\(sequence)", isDirectory: true)

        do {
            try FileManager.default.createDirectory(at: directory,
                                                    withIntermediateDirectories: true)
        } catch {
            log?("a directory for the arriving files could not be made: \(error)")
            return nil
        }

        var urls: [URL] = []
        var used: Set<String> = []
        var at = 0
        for file in delivery.files {
            let size = Int(file.size)
            guard at + size <= delivery.bytes.count else {
                log?("the arriving files did not add up to the payload; nothing was written")
                discard(directory)
                return nil
            }
            let name = FileNaming.unused(file.name, among: &used)
            let url = directory.appendingPathComponent(name)
            do {
                try Data(delivery.bytes[at..<(at + size)]).write(to: url, options: .atomic)
            } catch {
                log?("\(name) could not be written: \(error); the whole set was discarded")
                discard(directory)
                return nil
            }
            urls.append(url)
            at += size
        }

        guard !urls.isEmpty else {
            discard(directory)
            return nil
        }
        return Written(directory: directory, urls: urls)
    }

    private func discard(_ directory: URL) {
        try? FileManager.default.removeItem(at: directory)
    }
}

File: helpers/windows/src/file_store.cpp
#include "file_store.h"

#include <shlobj.h>

#include <cstring>
#include <cwchar>
#include <set>

#include "file_naming.h"

namespace deskhop {

namespace {

std::wstring utf8_to_wide(const std::string &text) {
    if (text.empty()) return {};
    const int chars = MultiByteToWideChar(CP_UTF8, 0, text.data(), static_cast<int>(text.size()),
                                          nullptr, 0);
    if (chars <= 0) return {};
    std::wstring out(static_cast<size_t>(chars), L'\0');
    MultiByteToWideChar(CP_UTF8, 0, text.data(), static_cast<int>(text.size()), out.data(), chars);
    return out;
}

/* Everything under `directory`, and the directory itself. `SHFileOperation`
   would do it in one call and is deliberately not used: it can put up UI, and
   a background helper must never do that. */
void remove_tree(const std::wstring &directory) {
    WIN32_FIND_DATAW found{};
    HANDLE search = FindFirstFileW((directory + L"\\*").c_str(), &found);
    if (search != INVALID_HANDLE_VALUE) {
        do {
            const std::wstring name = found.cFileName;
            if (name == L"." || name == L"..") continue;
            const std::wstring path = directory + L"\\" + name;
            if ((found.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0) remove_tree(path);
            else DeleteFileW(path.c_str());
        } while (FindNextFileW(search, &found));
        FindClose(search);
    }
    RemoveDirectoryW(directory.c_str());
}

} // namespace

FileStore::FileStore(std::wstring root) : root_(std::move(root)) {}

std::wstring FileStore::default_root() {
    wchar_t buffer[MAX_PATH]{};
    const DWORD length = GetTempPathW(MAX_PATH, buffer);
    if (length == 0 || length >= MAX_PATH) return L"deskhopplus";
    std::wstring path(buffer, length);
    if (!path.empty() && path.back() != L'\\') path += L'\\';
    return path + L"deskhopplus";
}

void FileStore::collect_garbage() {
    if (GetFileAttributesW(root_.c_str()) == INVALID_FILE_ATTRIBUTES) return;

    /* Newest by the counter in the name, which is what `write` builds it from:
       the tick count, then a per-run sequence. Compared as numbers rather than
       as text, so "10" sorts after "9". */
    std::vector<std::wstring> sets;
    WIN32_FIND_DATAW found{};
    HANDLE search = FindFirstFileW((root_ + L"\\*").c_str(), &found);
    if (search == INVALID_HANDLE_VALUE) return;
    do {
        const std::wstring name = found.cFileName;
        if (name == L"." || name == L"..") continue;
        sets.push_back(name);
    } while (FindNextFileW(search, &found));
    FindClose(search);

    std::wstring newest;
    unsigned long long newest_order[2] = {0, 0};
    for (const std::wstring &name : sets) {
        unsigned long long order[2] = {0, 0};
        const size_t dash = name.find(L'-');
        order[0] = std::wcstoull(name.substr(0, dash).c_str(), nullptr, 10);
        if (dash != std::wstring::npos)
            order[1] = std::wcstoull(name.substr(dash + 1).c_str(), nullptr, 10);
        if (newest.empty() || order[0] > newest_order[0] ||
            (order[0] == newest_order[0] && order[1] > newest_order[1])) {
            newest = name;
            newest_order[0] = order[0];
            newest_order[1] = order[1];
        }
    }

    unsigned removed = 0;
    for (const std::wstring &name : sets) {
        if (name == newest) continue;
        remove_tree(root_ + L"\\" + name);
        removed++;
    }
    if (removed > 0 && log)
        log(std::to_string(removed) +
            " set(s) of files left by a previous run were removed; the newest was kept in case "
            "it is still on the clipboard");
}

bool FileStore::write(const FileDelivery &delivery, Written &out) {
    sequence_++;
    const std::wstring directory = root_ + L"\\" + std::to_wstring(GetTickCount64()) + L"-" +
                                   std::to_wstring(sequence_);
    if (SHCreateDirectoryExW(nullptr, directory.c_str(), nullptr) != ERROR_SUCCESS &&
        GetFileAttributesW(directory.c_str()) == INVALID_FILE_ATTRIBUTES) {
        if (log) log("a directory for the arriving files could not be made");
        return false;
    }

    std::vector<std::wstring> paths;
    std::set<std::string> used;
    size_t at = 0;
    for (const FileEntry &file : delivery.files) {
        const size_t size = static_cast<size_t>(file.size);
        if (at + size > delivery.bytes.size()) {
            if (log) log("the arriving files did not add up to the payload; nothing was written");
            remove_tree(directory);
            return false;
        }
        const std::wstring path = directory + L"\\" + utf8_to_wide(unused_file_name(file.name, used));
        HANDLE handle = CreateFileW(path.c_str(), GENERIC_WRITE, 0, nullptr, CREATE_ALWAYS,
                                    FILE_ATTRIBUTE_NORMAL, nullptr);
        if (handle == INVALID_HANDLE_VALUE) {
            if (log) log("a file that arrived could not be created; the whole set was discarded");
            remove_tree(directory);
            return false;
        }
        size_t left = size;
        const uint8_t *from = delivery.bytes.data() + at;
        bool ok = true;
        while (left > 0) {
            const DWORD ask = left > 0x100000u ? 0x100000u : static_cast<DWORD>(left);
            DWORD wrote = 0;
            if (!WriteFile(handle, from, ask, &wrote, nullptr) || wrote == 0) {
                ok = false;
                break;
            }
            from += wrote;
            left -= wrote;
        }
        CloseHandle(handle);
        if (!ok) {
            if (log) log("a file that arrived could not be written; the whole set was discarded");
            remove_tree(directory);
            return false;
        }
        paths.push_back(path);
        at += size;
    }

    if (paths.empty()) {
        remove_tree(directory);
        return false;
    }
    out.directory = directory;
    out.paths = std::move(paths);
    return true;
}

} // namespace deskhop
